Exceptions hierarchy

EmanueleMinotto opened the pull request:
Some exceptions aren't general exceptions but exceptions defined by Penny, so the final developer should be able to catch them if he/she wants to.

samsonasik commented:
:+1:

wdalmut commented on src/Dispatcher.php:
```diff
@@ -66,9 +63,10 @@ public function __invoke(RequestInterface $request)
                 return $routeInfo;
             default:
-                throw new Exception(null, 500);
+                throw new Exception\Exception(null, 500);
```
This should be more specific like `InvalidRouteException` isn't it?

wdalmut replied:
Or maybe just `RuntimeException`

wdalmut replied:
Or a `PennyException` maybe...

wdalmut commented on src/Exception/Exception.php:
```diff
@@ -0,0 +1,9 @@
+
+use Exception as BaseException;
+
+class Exception extends BaseException
```
With `RuntimeException` we can drop completely this class

wdalmut replied:
Or drop it in favor of `PennyException`

EmanueleMinotto replied:
I'm not fully sure about this, because https://github.com/pennyphp/penny/pull/97/files#diff-58980ac23a3a9a0b92abd9805eed1e78R37 isn't a runtime exception.

In this PR the `PennyException` is just an interface for all the exceptions defined here, to prevent SPL hierarchy changes and mark them a "thrown by Penny" exceptions.

What do you think @gianarb @wdalmut ?

wdalmut commented on src/Exception/RouteNotFound.php:
```diff
@@ -2,8 +2,6 @@
 
-use Exception;
-
 final class RouteNotFound extends Exception
```
Is not better with `RouteNotFoundException` in order to be specific with `Exception` ?

EmanueleMinotto committed:
small coding style fixes

EmanueleMinotto committed:
added exceptions hierarchy

EmanueleMinotto committed:
fixed conflicts

# Conflicts:
#	src/Route/RouteInfo.php
#	tests/Route/RouteInfoTest.php
#	tests/Utils/FastSymfonyDispatcher.php

gianarb commented:
Hello @EmanueleMinotto I am perplexed around this PR.
I don't see improvement to extends:
- \RuntimeException
- \Exception

Only to have Penny Exception..

RouteNotFound and MethodNotAllowed explicit a specific exception..

but Penny\Exception doesn't add nothing at a normal \Exception same stuff for Penny\RuntimeExpection

EmanueleMinotto commented:
For example: if the final developer implements the `Interop\Container\ContainerInterface` interface in one of their classes, and the `has('router')` method throws an exception, how do you know if it's a container problem or a Penny problem?

Currently only reading the exception message, but exceptions messages should only go up to the UI instead of be used in the application level, also is more easy to catch:

``` php
try {
    $app = new App($wrongContainer);
} catch (PennyException $e) {
    // add the router
} catch (FakerContainerException $e) {
    // change the container
} catch (\Exception $e) {
    // check the whole application
} finally {
    // WTF?
}
```

gianarb commented:
You can not catch exceptions by `Penny\App` because it implements an error handler and it doesn't trigger exceptions :+1: 

I don't know IMO `Penny\Exception` doesn't add any plus respect a \Exception. it is only another class to maintain..

wdalmut commented:
I know that i am always the 12th man... But if there are no one exception thrown by the `App` it self?

``` php
// super index.php clear file
$app = new App(...);
$app->run();
```

An the framework uses the **event-driven** strategy to collect errors internally?

``` php
class App
{
  // somewhere
  try {

  } catch (Penny\PageNotFoundException $e) {  
    $this->eventManager->trigger("404", ["exception" => $e, "request" => $request]); 
  } catch (\Exception $e) {
    $this->eventManager->trigger("exception", ["exception" => $e]);
  }
}
```

In that way a user can attach any listener for errors:

``` php
$eventManager->attach("exception", function ($event) {
  // exception rendering (twig, or whoops or anything else that a user wants...)
});
```

https://filp.github.io/whoops/

What do you think on that?

EmanueleMinotto commented:
Totally agree with you @wdalmut :+1:
Probably mine is a not required overengineering, anyway since an exception could be thrown in the `Penny\App` constructor ([ref](https://github.com/pennyphp/penny/blob/master/src/App.php#L36)) there will still be the problem: is the container implementation or Penny itself?

If all the the other exceptions will be managed internally than I'll work to reduce the number of changes in this PR.

gianarb commented:
> An the framework uses the event-driven strategy to collect errors internally?

Yes we catch all errors :)

```
$eventManager->attach("*", function ($event) {
    if ($e = $event->getException() instanceof Exception) {
        throw $e;
    }
});
```

You can use `whoops` in your implementation.. or we can add it into the penny-classic-app :)

@EmanueleMinotto I'm very happy to implement only one part of your PR at the moment :) 

`MethodNotAllowedException` :+1: yes
`PennyRuntimeExcpetion` and `PennyException` at the moment no.. :) what do you think? :)

EmanueleMinotto commented:
@gianarb sounds great to me :) I'll ping you once completed
